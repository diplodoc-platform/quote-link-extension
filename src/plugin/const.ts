export const ENV_FLAG_NAME = 'has-yfm-quote-link';

export const TokenType = {
    QuoteLink: 'yfm_quote-link',
    QuoteLinkOpen: 'yfm_quote-link_open',
    QuoteLinkClose: 'yfm_quote-link_close',
} as const;

export const ClassNames = {
    QuoteLink: 'yfm-quote-link',
} as const;

export const QUOTE_LINK_ATTR = 'data-quotelink';
