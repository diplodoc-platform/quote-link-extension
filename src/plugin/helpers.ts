import type MdIt from 'markdown-it';

const matchCloseToken = (tokens: MdIt.Token[], i: number) => {
    return tokens[i].type === 'blockquote_close';
};

const matchOpenToken = (tokens: MdIt.Token[], i: number) => {
    return tokens[i].type === 'blockquote_open';
};

export function matchBlockquote(tokens: MdIt.Token[], idx: number) {
    if (!matchOpenToken(tokens, idx)) {
        return null;
    }

    let level = 0;
    let i = idx + 1;
    while (i < tokens.length) {
        if (matchOpenToken(tokens, i)) {
            level++;
        } else if (matchCloseToken(tokens, i)) {
            if (level === 0) {
                return {
                    openToken: tokens[idx],
                    closeToken: tokens[i],
                    closeTokenIndex: i,
                };
            }
            level--;
        }

        i++;
    }

    return null;
}

export function matchLinkAtInlineStart(inlineToken: MdIt.Token) {
    if (inlineToken.type !== 'inline' || !inlineToken.children?.length) {
        return null;
    }

    const {children: tokens} = inlineToken;
    if (tokens[0].type !== 'link_open') {
        return null;
    }

    for (let i = 0; i < tokens.length; i++) {
        const token = tokens[i];
        if (token.type === 'link_close') {
            return {
                openToken: tokens[0],
                closeToken: token,
                closeTokenIndex: i,
            };
        }
    }

    return null;
}
