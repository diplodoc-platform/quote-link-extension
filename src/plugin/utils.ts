export function hidden<B extends Record<string | symbol, unknown>, F extends string | symbol, V>(
    box: B,
    field: F,
    value: V,
) {
    if (!(field in box)) {
        Object.defineProperty(box, field, {
            enumerable: false,
            value: value,
        });
    }

    return box as B & {[P in F]: V};
}

export type Runtime = {
    script: string;
    style: string;
};

declare const __dirname: string;
export function copyRuntime(
    {runtime, output}: {runtime: Runtime; output: string},
    cache: Set<string>,
) {
    const PATH_TO_RUNTIME = '../runtime';
    const {join, resolve} = dynrequire('node:path');
    const runtimeFiles = {
        'index.js': runtime.script,
        'index.css': runtime.style,
    };
    for (const [originFile, outputFile] of Object.entries(runtimeFiles)) {
        const file = join(PATH_TO_RUNTIME, originFile);
        if (!cache.has(file)) {
            cache.add(file);
            copy(resolve(__dirname, file), join(output, outputFile));
        }
    }
}

export function copy(from: string, to: string) {
    const {mkdirSync, copyFileSync} = dynrequire('node:fs');
    const {dirname} = dynrequire('node:path');

    mkdirSync(dirname(to), {recursive: true});
    copyFileSync(from, to);
}

/*
 * Runtime require hidden for builders.
 * Used for nodejs api
 */
export function dynrequire(module: string) {
    // eslint-disable-next-line no-eval
    return eval(`require('${module}')`);
}
