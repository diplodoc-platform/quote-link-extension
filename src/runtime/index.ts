import './styles/quote-link.scss';
