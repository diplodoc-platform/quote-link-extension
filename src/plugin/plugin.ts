import type MarkdownIt from 'markdown-it';
import type Core from 'markdown-it/lib/parser_core';

import {parseMdAttrs} from '@diplodoc/utils';

import {ClassNames, ENV_FLAG_NAME, QUOTE_LINK_ATTR, TokenType} from './const';
import {matchBlockquote, matchLinkAtInlineStart} from './helpers';

export const quoteLinkPlugin: MarkdownIt.PluginSimple = (md) => {
    const plugin: Core.RuleCore = (state) => {
        const {tokens} = state;

        for (let i = 0; i < tokens.length; i++) {
            const quoteMatch = matchBlockquote(tokens, i);
            if (!quoteMatch) {
                continue;
            }

            if (tokens[i + 1]?.type !== 'paragraph_open' || !tokens[i + 2]) {
                continue;
            }

            const inlineToken = tokens[i + 2];
            const linkMatch = matchLinkAtInlineStart(inlineToken);
            if (!linkMatch) {
                continue;
            }

            // if attr not exist, parse markup after link
            if (linkMatch.openToken.attrIndex(QUOTE_LINK_ATTR) === -1) {
                const nextTextToken = inlineToken.children?.[linkMatch.closeTokenIndex + 1];
                if (!nextTextToken || nextTextToken.type !== 'text') {
                    continue;
                }

                const res = parseMdAttrs(
                    md,
                    nextTextToken.content,
                    0,
                    nextTextToken.content.length,
                );

                if (!res) {
                    continue;
                }

                nextTextToken.content = nextTextToken.content.slice(res.pos);

                if (res.attrs[QUOTE_LINK_ATTR]?.length) {
                    linkMatch.openToken.attrSet(QUOTE_LINK_ATTR, res.attrs[QUOTE_LINK_ATTR][0]);
                }
            }

            if (linkMatch.openToken.attrIndex(QUOTE_LINK_ATTR) !== -1) {
                quoteMatch.openToken.type = TokenType.QuoteLinkOpen;
                quoteMatch.openToken.attrSet('class', ClassNames.QuoteLink);

                quoteMatch.closeToken.type = TokenType.QuoteLinkClose;

                state.env ??= {};
                state.env[ENV_FLAG_NAME] = true;
            }
        }
    };

    md.core.ruler.push('quoteLink', plugin);
};
