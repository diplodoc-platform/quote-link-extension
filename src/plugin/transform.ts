import type MarkdownIt from 'markdown-it';

import {quoteLinkPlugin} from './plugin';
import {type Runtime, copyRuntime, dynrequire, hidden} from './utils';
import {ENV_FLAG_NAME} from './const';

export type TransformOptions = {
    runtime?:
        | string
        | {
              script: string;
              style: string;
          };
    bundle?: boolean;
};

type NormalizedPluginOptions = Omit<TransformOptions, 'runtime'> & {
    runtime: Runtime;
};

const registerTransform = (
    md: MarkdownIt,
    {
        runtime,
        bundle,
        output,
    }: Pick<NormalizedPluginOptions, 'bundle' | 'runtime'> & {
        output: string;
    },
) => {
    md.use(quoteLinkPlugin);

    md.core.ruler.push('yfm_quote-link_after', ({env}) => {
        hidden(env, 'bundled', new Set<string>());

        if (env?.[ENV_FLAG_NAME]) {
            env.meta = env.meta || {};
            env.meta.script = env.meta.script || [];
            env.meta.script.push(runtime.script);
            env.meta.style = env.meta.style || [];
            env.meta.style.push(runtime.style);

            if (bundle) {
                copyRuntime({runtime, output}, env.bundled);
            }
        }
    });
};

type InputOptions = {
    destRoot: string;
};

export function transform(options: Partial<TransformOptions> = {}) {
    const {bundle = true} = options;

    if (bundle && typeof options.runtime === 'string') {
        throw new TypeError('Option `runtime` should be record when `bundle` is enabled.');
    }

    const runtime: Runtime =
        typeof options.runtime === 'string'
            ? {script: options.runtime, style: options.runtime}
            : options.runtime || {
                  script: '_assets/quote-link-extension.js',
                  style: '_assets/quote-link-extension.css',
              };

    const plugin: MarkdownIt.PluginWithOptions<{output?: string}> = function (
        md: MarkdownIt,
        {output = '.'} = {},
    ) {
        registerTransform(md, {
            runtime,
            bundle,
            output,
        });
    };

    Object.assign(plugin, {
        collect(input: string, {destRoot = '.'}: InputOptions) {
            const MdIt = dynrequire('markdown-it');
            const md = new MdIt().use((md: MarkdownIt) => {
                registerTransform(md, {
                    runtime,
                    bundle,
                    output: destRoot,
                });
            });

            md.parse(input, {});
        },
    });

    return plugin;
}
