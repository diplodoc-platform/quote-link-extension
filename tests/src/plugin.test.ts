import transform from '@diplodoc/transform';
import dd from 'ts-dedent';
import MarkdownIt from 'markdown-it';
import attrs from 'markdown-it-attrs';

import * as quoteLinkExtension from '../../src/plugin';

const html = (text: string, opts?: quoteLinkExtension.TransformOptions) => {
    const {result} = transform(text, {
        plugins: [quoteLinkExtension.transform({bundle: false, ...opts})],
    });

    return result.html;
};

const meta = (text: string, opts?: quoteLinkExtension.TransformOptions) => {
    const {result} = transform(text, {
        plugins: [quoteLinkExtension.transform({bundle: false, ...opts})],
    });

    return result.meta;
};

const parse = (text: string, opts?: quoteLinkExtension.TransformOptions) => {
    const md = new MarkdownIt().use(quoteLinkExtension.transform({bundle: false, ...opts}));
    md.use(attrs, {leftDelimiter: '{', rightDelimiter: '}'});
    return md.parse(text, {});
};

describe('Quote link extension - plugin', () => {
    it('should render single quote link', () => {
        expect(
            html(dd`
            > [Quote link](https://ya.ru){data-quotelink}
            >
            > quote link text
            `),
        ).toBe(
            dd`
            <blockquote class="yfm-quote-link">
            <p><a href="https://ya.ru" data-quotelink>Quote link</a></p>
            <p>quote link text</p>
            </blockquote>
            ` + '\n',
        );
    });

    it('should render nested quote link', () => {
        expect(
            html(dd`
            > [Quote link](https://ya.ru){data-quotelink="true"}
            >
            > quote link text
            >
            > > [Nested](https://nested.ru){data-quotelink="true"}
            > >
            > > nested link text
            `),
        ).toBe(
            dd`
            <blockquote class="yfm-quote-link">
            <p><a href="https://ya.ru" data-quotelink="true">Quote link</a></p>
            <p>quote link text</p>
            <blockquote class="yfm-quote-link">
            <p><a href="https://nested.ru" data-quotelink="true">Nested</a></p>
            <p>nested link text</p>
            </blockquote>
            </blockquote>
            ` + '\n',
        );
    });

    it('should render simple quote inside quote link', () => {
        expect(
            html(dd`
            > [Quote link](https://ya.ru){data-quotelink="true"}
            >
            > quote link text
            >
            > > Simple quote
            > >
            > > simple quote text
            `),
        ).toBe(
            dd`
            <blockquote class="yfm-quote-link">
            <p><a href="https://ya.ru" data-quotelink="true">Quote link</a></p>
            <p>quote link text</p>
            <blockquote>
            <p>Simple quote</p>
            <p>simple quote text</p>
            </blockquote>
            </blockquote>
            ` + '\n',
        );

        expect(
            html(dd`
            > [Quote link](https://ya.ru){data-quotelink="true"}
            >
            > quote link text
            >
            > > [Simple quote with link](https://nested.ru)
            > >
            > > simple quote text
            `),
        ).toBe(
            dd`
            <blockquote class="yfm-quote-link">
            <p><a href="https://ya.ru" data-quotelink="true">Quote link</a></p>
            <p>quote link text</p>
            <blockquote>
            <p><a href="https://nested.ru">Simple quote with link</a></p>
            <p>simple quote text</p>
            </blockquote>
            </blockquote>
            ` + '\n',
        );
    });

    it('should render quote link inside simple quote', () => {
        expect(
            html(dd`
            > Simple quote
            >
            > simple quote text
            >
            > > [Nested quote link](https://nested.ru){data-quotelink="true"}
            > >
            > > nested quote link text
            `),
        ).toBe(
            dd`
            <blockquote>
            <p>Simple quote</p>
            <p>simple quote text</p>
            <blockquote class="yfm-quote-link">
            <p><a href="https://nested.ru" data-quotelink="true">Nested quote link</a></p>
            <p>nested quote link text</p>
            </blockquote>
            </blockquote>
            ` + '\n',
        );

        expect(
            html(dd`
            > [Simple link](https://ya.ru)
            >
            > simple link text
            >
            > > [Nested quote link](https://nested.ru){data-quotelink="true"}
            > >
            > > nested quote link text
            `),
        ).toBe(
            dd`
            <blockquote>
            <p><a href="https://ya.ru">Simple link</a></p>
            <p>simple link text</p>
            <blockquote class="yfm-quote-link">
            <p><a href="https://nested.ru" data-quotelink="true">Nested quote link</a></p>
            <p>nested quote link text</p>
            </blockquote>
            </blockquote>
            ` + '\n',
        );
    });

    it('should render simple quote without link', () => {
        expect(html('> quote text')).toBe(
            dd`
            <blockquote>
            <p>quote text</p>
            </blockquote>
            ` + '\n',
        );

        expect(
            html(dd`
        > [Simple link](https://ya.ru)
        >
        > link text
        `),
        ).toBe(
            dd`
            <blockquote>
            <p><a href="https://ya.ru">Simple link</a></p>
            <p>link text</p>
            </blockquote>
            ` + '\n',
        );
    });

    it('should not add assets to meta if no yfm-quote-link is found', () => {
        expect(meta('paragraph')).toBeUndefined();
    });

    it('should add default assets to meta', () => {
        expect(
            meta(dd`
            > [Quote link](https://ya.ru){data-quotelink="true"}
            >
            > quote link text
            `),
        ).toStrictEqual({
            script: ['_assets/quote-link-extension.js'],
            style: ['_assets/quote-link-extension.css'],
        });
    });

    it('should add custom assets to meta', () => {
        expect(
            meta(
                dd`
                > [Quote link](https://ya.ru){data-quotelink="true"}
                >
                > quote link text
                `,
                {runtime: 'yfm-quote-link'},
            ),
        ).toStrictEqual({
            script: ['yfm-quote-link'],
            style: ['yfm-quote-link'],
        });
    });

    it('should add custom assets to meta 2', () => {
        expect(
            meta(
                dd`
                > [Quote link](https://ya.ru){data-quotelink="true"}
                >
                > quote link text
                `,
                {
                    runtime: {script: 'yfm-quote-link.script', style: 'yfm-quote-link.style'},
                },
            ),
        ).toStrictEqual({script: ['yfm-quote-link.script'], style: ['yfm-quote-link.style']});
    });

    it('should parse markup with quote link to token stream', () => {
        expect(
            parse(dd`
            > [Quote link](https://ya.ru){data-quotelink="true"} content after qoute link
            >
            > quote link text
            `),
        ).toMatchSnapshot();
    });
});
