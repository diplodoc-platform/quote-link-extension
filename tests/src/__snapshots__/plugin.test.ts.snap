// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quote link extension - plugin should parse markup with quote link to token stream 1`] = `
[
  Token {
    "attrs": [
      [
        "class",
        "yfm-quote-link",
      ],
    ],
    "block": true,
    "children": null,
    "content": "",
    "hidden": false,
    "info": "",
    "level": 0,
    "map": [
      0,
      3,
    ],
    "markup": ">",
    "meta": null,
    "nesting": 1,
    "tag": "blockquote",
    "type": "yfm_quote-link_open",
  },
  Token {
    "attrs": null,
    "block": true,
    "children": null,
    "content": "",
    "hidden": false,
    "info": "",
    "level": 1,
    "map": [
      0,
      1,
    ],
    "markup": "",
    "meta": null,
    "nesting": 1,
    "tag": "p",
    "type": "paragraph_open",
  },
  Token {
    "attrs": null,
    "block": true,
    "children": [
      Token {
        "attrs": [
          [
            "href",
            "https://ya.ru",
          ],
          [
            "data-quotelink",
            "true",
          ],
        ],
        "block": false,
        "children": null,
        "content": "",
        "hidden": false,
        "info": "",
        "level": 0,
        "map": null,
        "markup": "",
        "meta": null,
        "nesting": 1,
        "tag": "a",
        "type": "link_open",
      },
      Token {
        "attrs": null,
        "block": false,
        "children": null,
        "content": "Quote link",
        "hidden": false,
        "info": "",
        "level": 1,
        "map": null,
        "markup": "",
        "meta": null,
        "nesting": 0,
        "tag": "",
        "type": "text",
      },
      Token {
        "attrs": null,
        "block": false,
        "children": null,
        "content": "",
        "hidden": false,
        "info": "",
        "level": 0,
        "map": null,
        "markup": "",
        "meta": null,
        "nesting": -1,
        "tag": "a",
        "type": "link_close",
      },
      Token {
        "attrs": null,
        "block": false,
        "children": null,
        "content": " content after qoute link",
        "hidden": false,
        "info": "",
        "level": 0,
        "map": null,
        "markup": "",
        "meta": null,
        "nesting": 0,
        "tag": "",
        "type": "text",
      },
    ],
    "content": "[Quote link](https://ya.ru){data-quotelink="true"} content after qoute link",
    "hidden": false,
    "info": "",
    "level": 2,
    "map": [
      0,
      1,
    ],
    "markup": "",
    "meta": null,
    "nesting": 0,
    "tag": "",
    "type": "inline",
  },
  Token {
    "attrs": null,
    "block": true,
    "children": null,
    "content": "",
    "hidden": false,
    "info": "",
    "level": 1,
    "map": null,
    "markup": "",
    "meta": null,
    "nesting": -1,
    "tag": "p",
    "type": "paragraph_close",
  },
  Token {
    "attrs": null,
    "block": true,
    "children": null,
    "content": "",
    "hidden": false,
    "info": "",
    "level": 1,
    "map": [
      2,
      3,
    ],
    "markup": "",
    "meta": null,
    "nesting": 1,
    "tag": "p",
    "type": "paragraph_open",
  },
  Token {
    "attrs": null,
    "block": true,
    "children": [
      Token {
        "attrs": null,
        "block": false,
        "children": null,
        "content": "quote link text",
        "hidden": false,
        "info": "",
        "level": 0,
        "map": null,
        "markup": "",
        "meta": null,
        "nesting": 0,
        "tag": "",
        "type": "text",
      },
    ],
    "content": "quote link text",
    "hidden": false,
    "info": "",
    "level": 2,
    "map": [
      2,
      3,
    ],
    "markup": "",
    "meta": null,
    "nesting": 0,
    "tag": "",
    "type": "inline",
  },
  Token {
    "attrs": null,
    "block": true,
    "children": null,
    "content": "",
    "hidden": false,
    "info": "",
    "level": 1,
    "map": null,
    "markup": "",
    "meta": null,
    "nesting": -1,
    "tag": "p",
    "type": "paragraph_close",
  },
  Token {
    "attrs": null,
    "block": true,
    "children": null,
    "content": "",
    "hidden": false,
    "info": "",
    "level": 0,
    "map": null,
    "markup": ">",
    "meta": null,
    "nesting": -1,
    "tag": "blockquote",
    "type": "yfm_quote-link_close",
  },
]
`;
